### DISCUSSION TITLE
Execute Shell Commands From MOOSE

### DISCUSSION THREAD

**Discussion by josebastiase** [idaholab/moose#25157] [Q&A General]
Hi all,
I am interested in running a Python script that performs calculations using simulation results after each time step. Subsequently, this Python script generates a CSV file, which in turn adjusts the boundary conditions within my simulation.
As the Python script can be executed via a terminal, I am wondering if it is currently possible to run execute shell commands from MOOSE.
Thanks in advance!
Jose

**Comment by GiudGiud** [comment]
Hello

We unfortunately dont have this capability right now but it could be done in a user object with a bit of C++.
And we'd like to have this one day

@loganharbour for confirmation we dont have this

Guillaume

**Reply by loganharbour** [reply]
We don't have the capability. Execution paths that are introduced like this can be very difficult to test, therefore in most cases we'd rather develop a more framework-esque solution.

Something like this will coming soon for a specific funding need we have, but I'm still weary to using it in most cases

**Reply by GiudGiud** [reply]
I d be fine with a RunCommandUO. 
we can impose some constraints on what commands once we see what issues appear

**Reply by josebastiase** [reply]
Hi,
 Can I attempt to write C++ code within a user object based on popen? Would that be effective?
Something like:

```

std::string exec(const char* cmd) {
   
    std::array<char, 128> buffer;
    std::string result;
    std::unique_ptr<FILE, decltype(&pclose)> pipe(popen(cmd, "r"), pclose);
    if (!pipe) {
        throw std::runtime_error("popen() failed");
    }
    
    while (fgets(buffer.data(), buffer.size(), pipe.get()) != nullptr) {
        result += buffer.data();
    }
    
    return result;
}
```

**Reply by GiudGiud** [reply]
Not sure if really need to capture pipe output. I guess if that's what you expect from your script it's fine.

**Reply by josebastiase** [reply]
Hi @GiudGiud,

my pipe between MOOSE and the shell appears to be functioning correctly. However, I am encountering issues when attempting to run it in parallel. The `UserObject `is being instantiated the same number of times as the number of cores I use for my MOOSE simulation. How can I ensure that it is only called once? 
I am using a `GeneralUserObject`
thanks!

**Reply by GiudGiud** [reply]
Hello

You can't avoid the construction of the objects on every process.
You can prevent a line / function from being called by doing:
```
if (processor_id() == 0)
{
...
}
```

**Comment by veeshy** [comment]
I've used [pybind11](https://github.com/pybind/pybind11) + user object for this, it's not trivial to setup but no real gotchas either.

**Reply by josebastiase** [reply]
Hi,
thanks, great idea! Are you willing to share your code?
